`df`: Adds support for mount path prefix matching and input path

crazystylus opened the pull request:
Adds support for mount path prefix matching and input path canonicalization.
Fixes #3065

- Sorts mount paths in the order of decreasing string length
- Canonicalize all paths and clear invalid paths
- Checking of mount path prefix matches input path

crazystylus commented:
I don't think the failed CI/CD job has something to do with the code as the logs mentions failure of apt install `Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/n/npm/npm_6.14.4+ds-1ubuntu2_all.deb  Unable to connect to azure.archive.ubuntu.com:http: [IP: 52.250.76.244 80]`

jfinkels commented on src/uu/df/src/df.rs:
```diff
@@ -274,8 +298,11 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+    let mut mounts = read_fs_list();
+
+    // Sort mounts in desc ordered lexicographically
+    mounts.sort_by(|a, b| b.mount_dir.cmp(&a.mount_dir));
```
Why sort here? On my system (Ubuntu), the output of `df` is not lexicographically ordered:

```
$ df --output=source | tail -n +2 | head -n4
udev
tmpfs
/dev/mapper/vgubuntu-root
tmpfs
```

See also my comment here: https://github.com/uutils/coreutils/issues/3086#issuecomment-1030965538

crazystylus replied:
My plan was to sort in the order of decreasing path length, so that when I do a prefix matching, the longest matching mount path is selected first and rest paths are ignored. For example if we have mount paths [`\`,`\root`] and we have input `/root/docs` then we only need to output `/root` and `/` needs to be skipped. Sorting helps to assure this part

Let me update the sorting section and remove lexicographical part and change to sort by length.

crazystylus commented:
I have made changes to mount_paths in order of decreasing length from reverse lexicographical order

jfinkels reviewed (dismissed):
This will cause a regression in `df` as it is today, due to the sorting. Unfortunately there is no unit test to cover this case, but the output of `df` should match the order of files in `/etc/mtab`, as I mentioned in this comment: https://github.com/uutils/coreutils/issues/3086#issuecomment-1030965538

Also, the various mutations of `paths` and the use of the empty string as a way of filtering out paths are tough to follow. Is it possible to move some of this logic into the `mount_info_lt()` function, which returns `true` exactly when one `MountInfo` object should be kept and the other dropped?

crazystylus commented:
> This will cause a regression in `df` as it is today, due to the sorting. Unfortunately there is no unit test to cover this case, but the output of `df` should match the order of files in `/etc/mtab`, as I mentioned in this comment: [#3086 (comment)](https://github.com/uutils/coreutils/issues/3086#issuecomment-1030965538)
> 
> Also, the various mutations of `paths` and the use of the empty string as a way of filtering out paths are tough to follow. Is it possible to move some of this logic into the `mount_info_lt()` function, which returns `true` exactly when one `MountInfo` object should be kept and the other dropped?

Can try running this and compare the output?
`df /boot / /boot/efi` and `df /boot/efi /boot /`
I think df will sort output according to input paths. If no path list is supplied, then only it will in the same order as /etc/mtab

crazystylus commented:
I have made changes as below for cleanup of code and meeting requirements
- Removed sorting of mount_paths
- Moved the canonicalization logic so that no mutability is required
- Moved path matching logic from `is_included()` to `filter_mount_list()`
- Changed loops to iterators to cleanup code where ever possible
- For each input_path, code selects longest mount_path and de-duplication is `is_best()`
- Removed string clear and mutability

jfinkels commented:
> Can try running this and compare the output?
> `df /boot / /boot/efi` and `df /boot/efi /boot /`
> I think df will order output according to input paths. If no path list is supplied, then only it will in the same order as /etc/mtab

Oh you are exactly right:
```
$ df --output=target /boot / /boot/efi
Mounted on
/boot
/
/boot/efi
$ df --output=target / /boot /boot/efi
Mounted on
/
/boot
/boot/efi
```
Sorry for my misunderstanding earlier. In that case, can you add a test for this in `tests/by-util/test_df.rs`?

crazystylus commented:
I have verified that `df` from main branch is falsely passing `test: df/df-symlink.sh` by giving no output because this test case passes `.` as input.

crazystylus commented:
Hi @jfinkels ,
I realized there is one more issue. We should be selecting best mount point per input_path, and if the input path is repeated n times, output should also contain n entries.
Try this to verify
```shell
$ df --output=target /boot /boot /boot
Mounted on
/boot
/boot
/boot
```

crazystylus commented:
I have checked GNU coreutils df code, and adjusted the helper functions according to it.
1. Mount list is only filtered if there are no input_paths and all entries are emitted
2. Else each input_path is canonicalized and then mapped with a mount_entry from unfiltered_list

Ref: https://github.com/coreutils/coreutils/blob/master/src/df.c#L1830-L1840

jfinkels commented:
> I have checked GNU coreutils df code, and adjusted the helper functions according to it.

It's best not to view the source code of GNU coreutils in order to avoid infringing their copyright.

>     1. Mount list is only filtered if there are no input_paths and all entries are emitted
>     2. Else each input_path is canonicalized and then mapped with a mount_entry from unfiltered_list

There have been a lot of changes in this pull request and it's hard for me to interpret. Could you give concrete examples of how the behavior of `df` is changing after these two changes?

> Both test cases have been added

I only see one new test case, asserting that if two files are specified in the input then two identical rows appear in the output table.

crazystylus commented:
> > I have checked GNU coreutils df code, and adjusted the helper functions according to it.
> 
> It's best not to view the source code of GNU coreutils in order to avoid infringing their copyright.
> 
> > ```
> > 1. Mount list is only filtered if there are no input_paths and all entries are emitted
> > 2. Else each input_path is canonicalized and then mapped with a mount_entry from unfiltered_list
> > ```
> 
> There have been a lot of changes in this pull request and it's hard for me to interpret. Could you give concrete examples of how the behavior of `df` is changing after these two changes?
> 
> > Both test cases have been added
> 
> I only see one new test case, asserting that if two files are specified in the input then two identical rows appear in the output table.

I had to remove the 2nd test case as it will never pass in the CI/CD Pipeline because only on mount point `/` is available for testing.

The following are the changes made by this PR:
1. All input paths are canonicalized, so relative paths and symlinks will start working now.
E.g. `df .` `df ../` df `symlink` will work successfully
2. Mount points are matched to input_path by prefix matching and longest one is selected
E.g. `df /root/....../long_path/....` will point to correct mount_point
3. Mount points are printed in the order of input paths
E.g. `df / /boot /boot/efi` `df /boot/efi /boot /` will have different output
4. If input path is repeated, it's subsequent matched entry is also repeated
E.g. `df / / /` will print 3 entries
5. If no paths is passed as argument to `df` then, filtered mount list is printed preserving the mtab order
E.g. `df`
6. GNU `Test: df/df-symlink.sh` is evaluated properly. Main branch code falsely passes by printing no output.
7. `df` will start passing GNU `Test: df/df-symlink.sh` when this PR and #3176 are both merged to main branch

jfinkels commented on src/uu/df/src/df.rs:
```diff
@@ -275,37 +268,59 @@ fn is_best(previous: &[MountInfo], mi: &MountInfo) -> bool {
 }
 
+/// Assign 1 `MountInfo` entry to each path
+/// `lofs` enries are skipped and dummy mount points are skipped
```
What does "lofs" mean?

crazystylus replied:
Here `lofs` is for Solaris style `loopback filesystem` which is not present in linux. It's present in Solaris and FreeBSD and is similar to symlink.

crazystylus commented:
I had to remove other test case also as it's only working in Linux and fails in Windows. Hence I think unit test cases is the best approach.

crazystylus commented:
Please let me know if any other changes are required in this PR for getting it merged.
